File: project3/backend/_archived/api-gateway-central-hub-shared-duplicate/js/utils/BaseService.js
/**
 * JavaScript BaseService - Compatible dengan Python BaseService
 * Provides standard service patterns untuk JavaScript services
 * Matches interface dari utils/base_service.py
 */

const EventEmitter = require('events');

class ServiceConfig {
    constructor({
        service_name,
        version,
        port,
        environment = 'development',
        enable_tracing = true,
        enable_circuit_breaker = true,
        health_check_interval = 30,
        cache_ttl_default = 300,
        max_connections = 100
    }) {
        this.service_name = service_name;
        this.version = version;
        this.port = port;
        this.environment = environment;
        this.enable_tracing = enable_tracing;
        this.enable_circuit_breaker = enable_circuit_breaker;
        this.health_check_interval = health_check_interval;
        this.cache_ttl_default = cache_ttl_default;
        this.max_connections = max_connections;
    }
}

class StandardResponse {
    constructor({
        success,
        data = null,
        error_message = null,
        processing_time_ms = 0,
        correlation_id = null
    }) {
        this.success = success;
        this.data = data;
        this.error_message = error_message;
        this.processing_time_ms = processing_time_ms;
        this.correlation_id = correlation_id;
        this.timestamp = new Date().toISOString();
    }
}

class BaseService extends EventEmitter {
    constructor(config) {
        super();

        if (!(config instanceof ServiceConfig)) {
            throw new Error('Config must be instance of ServiceConfig');
        }

        this.config = config;
        this.service_name = config.service_name;
        this.version = config.version;
        this.start_time = new Date();

        // Service state
        this.is_healthy = true;
        this.active_connections = 0;
        this.request_count = 0;
        this.error_count = 0;

        // Initialize components
        this._initializeComponents();

        // Setup logging
        this.logger = this._setupLogging();

        // Performance tracking
        this.response_times = [];
        this.request_timestamps = [];

        // Health check scheduler
        this._health_check_timer = null;
    }

    _initializeComponents() {
        // Initialize standard components (akan di-implement)
        this.db = null;           // Database manager
        this.cache = null;        // Cache manager
        this.config_manager = null; // Config manager
        this.tracer = null;       // Request tracer
        this.circuit_breaker = null; // Circuit breaker

        // Placeholder implementations
        if (this.config.enable_tracing) {
            this.tracer = {
                trace: (operation_name, correlation_id) => {
                    return {
                        end: () => {},
                        addError: (error) => {}
                    };
                }
            };
        }

        if (this.config.enable_circuit_breaker) {
            this.circuit_breaker = {
                is_open: (service) => false,
                record_success: async (service) => {},
                record_failure: async (service) => {},
                get_status: async () => ({ status: 'closed' })
            };
        }
    }

    _setupLogging() {
        // Use existing log_utils or create structured logger
        const logger = {
            info: (message, extra = {}) => {
                console.log(JSON.stringify({
                    timestamp: new Date().toISOString(),
                    level: 'INFO',
                    service: this.service_name,
                    message,
                    ...extra
                }));
            },
            warn: (message, extra = {}) => {
                console.warn(JSON.stringify({
                    timestamp: new Date().toISOString(),
                    level: 'WARN',
                    service: this.service_name,
                    message,
                    ...extra
                }));
            },
            error: (message, extra = {}) => {
                console.error(JSON.stringify({
                    timestamp: new Date().toISOString(),
                    level: 'ERROR',
                    service: this.service_name,
                    message,
                    ...extra
                }));
            },
            debug: (message, extra = {}) => {
                if (process.env.NODE_ENV === 'development') {
                    console.debug(JSON.stringify({
                        timestamp: new Date().toISOString(),
                        level: 'DEBUG',
                        service: this.service_name,
                        message,
                        ...extra
                    }));
                }
            }
        };

        return logger;
    }

    async processWithTracing(operation_name, func, correlation_id = null, ...args) {
        const start_time = Date.now();

        try {
            this.active_connections++;
            this.request_count++;

            let result;

            // Start tracing if enabled
            if (this.tracer && correlation_id) {
                const span = this.tracer.trace(operation_name, correlation_id);
                try {
                    result = await func(...args);
                } finally {
                    span.end();
                }
            } else {
                result = await func(...args);
            }

            // Calculate processing time
            const processing_time_ms = Date.now() - start_time;
            this.response_times.push(processing_time_ms);
            this.request_timestamps.push(Date.now());

            // Log success
            this.logger.info(`${operation_name} completed`, {
                operation: operation_name,
                processing_time_ms,
                correlation_id,
                success: true
            });

            return new StandardResponse({
                success: true,
                data: result,
                processing_time_ms,
                correlation_id
            });

        } catch (error) {
            this.error_count++;
            const processing_time_ms = Date.now() - start_time;

            // Log error
            this.logger.error(`${operation_name} failed: ${error.message}`, {
                operation: operation_name,
                processing_time_ms,
                correlation_id,
                error: error.message,
                success: false
            });

            // Add error to trace if tracing enabled
            if (this.tracer && correlation_id) {
                this.tracer.addError(correlation_id, error.message);
            }

            return new StandardResponse({
                success: false,
                error_message: error.message,
                processing_time_ms,
                correlation_id
            });

        } finally {
            this.active_connections--;
        }
    }

    async getConfig(key, defaultValue = null) {
        if (this.config_manager) {
            return await this.config_manager.get(key, defaultValue);
        }
        return process.env[key] || defaultValue;
    }

    async cacheGet(key, defaultValue = null) {
        if (this.cache) {
            return await this.cache.get(key, defaultValue);
        }
        return defaultValue;
    }

    async cacheSet(key, value, ttl = null) {
        if (this.cache) {
            return await this.cache.set(key, value, ttl);
        }
        return false;
    }

    async dbExecute(query, params = null) {
        if (this.db) {
            return await this.db.execute(query, params);
        }
        throw new Error('Database manager not initialized');
    }

    async checkCircuitBreaker(external_service) {
        if (this.circuit_breaker) {
            return this.circuit_breaker.is_open(external_service);
        }
        return false;
    }

    async recordExternalSuccess(external_service) {
        if (this.circuit_breaker) {
            await this.circuit_breaker.record_success(external_service);
        }
    }

    async recordExternalFailure(external_service) {
        if (this.circuit_breaker) {
            await this.circuit_breaker.record_failure(external_service);
        }
    }

    async healthCheck() {
        const uptime_seconds = Math.floor((Date.now() - this.start_time.getTime()) / 1000);

        const health_status = {
            service: this.service_name,
            status: this.is_healthy ? 'healthy' : 'unhealthy',
            version: this.version,
            timestamp: new Date().toISOString(),
            uptime_seconds,
            active_connections: this.active_connections,
            total_requests: this.request_count,
            error_rate: this.request_count > 0 ? this.error_count / this.request_count : 0,
            avg_response_time_ms: this._getAverageResponseTime()
        };

        try {
            // Check database connection
            if (this.db && typeof this.db.health_check === 'function') {
                health_status.database = await this.db.health_check();
            }

            // Check cache connection
            if (this.cache && typeof this.cache.health_check === 'function') {
                health_status.cache = await this.cache.health_check();
            }

            // Check circuit breaker status
            if (this.circuit_breaker) {
                health_status.circuit_breakers = await this.circuit_breaker.get_status();
            }

            // Add service-specific health checks
            const custom_health = await this.customHealthChecks();
            if (custom_health) {
                Object.assign(health_status, custom_health);
            }

        } catch (error) {
            this.logger.error(`Health check failed: ${error.message}`);
            health_status.status = 'degraded';
            health_status.error = error.message;
            this.is_healthy = false;
        }

        return health_status;
    }

    _getAverageResponseTime() {
        if (this.response_times.length === 0) return 0;
        const sum = this.response_times.reduce((a, b) => a + b, 0);
        return Math.round(sum / this.response_times.length);
    }

    async customHealthChecks() {
        // Override in child classes
        return {};
    }

    async start() {
        this.logger.info(`Starting ${this.service_name} v${this.version}`);

        try {
            // Initialize connections
            if (this.db && typeof this.db.connect === 'function') {
                await this.db.connect();
            }

            if (this.cache && typeof this.cache.connect === 'function') {
                await this.cache.connect();
            }

            if (this.config_manager && typeof this.config_manager.load_config === 'function') {
                await this.config_manager.load_config();
            }

            // Start health check scheduler
            if (this.config.health_check_interval > 0) {
                this._startHealthCheckScheduler();
            }

            // Custom startup logic
            await this.onStartup();

            this.logger.info(`${this.service_name} started successfully`);
            this.emit('started');

        } catch (error) {
            this.logger.error(`Failed to start ${this.service_name}: ${error.message}`);
            throw error;
        }
    }

    async stop() {
        this.logger.info(`Stopping ${this.service_name}`);

        try {
            // Stop health check scheduler
            if (this._health_check_timer) {
                clearInterval(this._health_check_timer);
                this._health_check_timer = null;
            }

            // Custom shutdown logic
            await this.onShutdown();

            // Close connections
            if (this.db && typeof this.db.disconnect === 'function') {
                await this.db.disconnect();
            }

            if (this.cache && typeof this.cache.disconnect === 'function') {
                await this.cache.disconnect();
            }

            this.logger.info(`${this.service_name} stopped`);
            this.emit('stopped');

        } catch (error) {
            this.logger.error(`Error during shutdown: ${error.message}`);
        }
    }

    _startHealthCheckScheduler() {
        this._health_check_timer = setInterval(async () => {
            try {
                await this.healthCheck();
            } catch (error) {
                this.logger.error(`Health check scheduler error: ${error.message}`);
            }
        }, this.config.health_check_interval * 1000);
    }

    // Abstract methods - implement in child classes
    async onStartup() {
        // Custom startup logic
    }

    async onShutdown() {
        // Custom shutdown logic
    }

    generateCorrelationId() {
        return `${this.service_name}_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    }
}

module.exports = {
    BaseService,
    ServiceConfig,
    StandardResponse
};

File: project3/backend/01-core-infrastructure/central-hub/shared/components/utils/base_service.py
"""
Base Service Class untuk standardisasi internal patterns
Digunakan oleh semua backend services untuk konsistensi
"""

import time
import asyncio
import logging as python_logging
from typing import Any, Dict, Optional, Callable
from dataclasses import dataclass
from datetime import datetime
from abc import ABC, abstractmethod

from .patterns.database import DatabaseManager
from .patterns.cache import CacheManager
from .patterns.config import InfrastructureConfigManager
from .patterns.tracing import RequestTracer
from .patterns.circuit_breaker import CircuitBreakerManager
from .patterns.response import StandardResponse


@dataclass
class ServiceConfig:
    """Standard service configuration"""
    service_name: str
    version: str
    port: int
    environment: str = "development"
    enable_tracing: bool = True
    enable_circuit_breaker: bool = True
    health_check_interval: int = 30
    cache_ttl_default: int = 300
    max_connections: int = 100


class BaseService(ABC):
    """
    Base class untuk semua backend services
    Menyediakan standard patterns untuk:
    - Database access
    - Caching
    - Configuration management
    - Request tracing
    - Circuit breaker protection
    - Error handling
    - Performance monitoring
    """

    def __init__(self, config: ServiceConfig):
        self.config = config
        self.service_name = config.service_name
        self.version = config.version
        self.start_time = datetime.utcnow()

        # Initialize standard components
        self._initialize_components()

        # Setup logging
        self.logger = self._setup_logging()

        # Service state
        self.is_healthy = True
        self.active_connections = 0

    def _initialize_components(self):
        """Initialize standard service components"""

        # Database manager
        self.db = DatabaseManager(
            service_name=self.service_name,
            max_connections=self.config.max_connections
        )

        # Cache manager
        self.cache = CacheManager(
            service_name=self.service_name,
            default_ttl=self.config.cache_ttl_default
        )

        # Infrastructure configuration manager (for Central Hub internal use)
        self.config_manager = InfrastructureConfigManager(
            service_name=self.service_name,
            environment=self.config.environment
        )

        # Request tracer
        if self.config.enable_tracing:
            self.tracer = RequestTracer(
                service_name=self.service_name
            )

        # Circuit breaker
        if self.config.enable_circuit_breaker:
            self.circuit_breaker = CircuitBreakerManager(
                service_name=self.service_name
            )

    def _setup_logging(self) -> python_logging.Logger:
        """Setup structured logging with newline format"""
        logger = python_logging.getLogger(self.service_name)
        logger.setLevel(python_logging.INFO)

        # Structured formatter with newline for better readability
        formatter = python_logging.Formatter(
            '%(asctime)s - %(name)s\n%(levelname)s - %(message)s'
        )

        handler = python_logging.StreamHandler()
        handler.setFormatter(formatter)
        logger.addHandler(handler)

        return logger

    async def process_with_tracing(self,
                                 operation_name: str,
                                 func: Callable,
                                 correlation_id: Optional[str] = None,
                                 *args, **kwargs) -> StandardResponse:
        """
        Standard wrapper untuk processing dengan tracing dan error handling
        """
        start_time = time.time()

        try:
            # Start tracing if enabled
            if hasattr(self, 'tracer') and correlation_id:
                with self.tracer.trace(operation_name, correlation_id):
                    result = await func(*args, **kwargs)
            else:
                result = await func(*args, **kwargs)

            # Calculate processing time
            processing_time_ms = (time.time() - start_time) * 1000

            # Log success
            self.logger.info(f"{operation_name} completed", extra={
                "operation": operation_name,
                "processing_time_ms": processing_time_ms,
                "correlation_id": correlation_id,
                "success": True
            })

            return StandardResponse(
                success=True,
                data=result,
                processing_time_ms=processing_time_ms,
                correlation_id=correlation_id
            )

        except Exception as e:
            # Calculate processing time
            processing_time_ms = (time.time() - start_time) * 1000

            # Log error
            self.logger.error(f"{operation_name} failed: {str(e)}", extra={
                "operation": operation_name,
                "processing_time_ms": processing_time_ms,
                "correlation_id": correlation_id,
                "error": str(e),
                "success": False
            })

            # Add error to trace if tracing enabled
            if hasattr(self, 'tracer') and correlation_id:
                self.tracer.add_error(correlation_id, str(e))

            return StandardResponse(
                success=False,
                error_message=str(e),
                processing_time_ms=processing_time_ms,
                correlation_id=correlation_id
            )

    async def get_config(self, key: str, default: Any = None) -> Any:
        """Standard configuration access"""
        return await self.config_manager.get(key, default)

    async def cache_get(self, key: str, default: Any = None) -> Any:
        """Standard cache access"""
        return await self.cache.get(key, default)

    async def cache_set(self, key: str, value: Any, ttl: Optional[int] = None) -> bool:
        """Standard cache storage"""
        return await self.cache.set(key, value, ttl)

    async def db_execute(self, query: str, params: Optional[Dict] = None) -> Any:
        """Standard database execution"""
        return await self.db.execute(query, params)

    async def check_circuit_breaker(self, external_service: str) -> bool:
        """Check if circuit breaker is open for external service"""
        if hasattr(self, 'circuit_breaker'):
            return self.circuit_breaker.is_open(external_service)
        return False

    async def record_external_success(self, external_service: str):
        """Record successful external service call"""
        if hasattr(self, 'circuit_breaker'):
            await self.circuit_breaker.record_success(external_service)

    async def record_external_failure(self, external_service: str):
        """Record failed external service call"""
        if hasattr(self, 'circuit_breaker'):
            await self.circuit_breaker.record_failure(external_service)

    async def health_check(self) -> Dict[str, Any]:
        """
        Standard health check implementation
        Override this method in child classes for service-specific checks
        """
        health_status = {
            "service": self.service_name,
            "status": "healthy" if self.is_healthy else "unhealthy",
            "version": self.version,
            "timestamp": datetime.utcnow().isoformat(),
            "uptime_seconds": int((datetime.utcnow() - self.start_time).total_seconds()),
            "active_connections": self.active_connections
        }

        try:
            # Check database connection
            if hasattr(self, 'db'):
                db_status = await self.db.health_check()
                health_status["database"] = db_status

            # Check cache connection
            if hasattr(self, 'cache'):
                cache_status = await self.cache.health_check()
                health_status["cache"] = cache_status

            # Check circuit breaker status
            if hasattr(self, 'circuit_breaker'):
                circuit_status = await self.circuit_breaker.get_status()
                health_status["circuit_breakers"] = circuit_status

            # Add service-specific health checks
            custom_health = await self.custom_health_checks()
            if custom_health:
                health_status.update(custom_health)

        except Exception as e:
            self.logger.error(f"Health check failed: {str(e)}")
            health_status["status"] = "degraded"
            health_status["error"] = str(e)
            self.is_healthy = False

        return health_status

    @abstractmethod
    async def custom_health_checks(self) -> Dict[str, Any]:
        """
        Service-specific health checks
        Implement this method in child classes
        """
        return {}

    async def start(self):
        """Start the service"""
        self.logger.info(f"Starting {self.service_name} v{self.version}")

        # Initialize connections
        await self.db.connect()
        await self.cache.connect()
        await self.config_manager.load_config()

        # Start health check scheduler
        if self.config.health_check_interval > 0:
            asyncio.create_task(self._health_check_scheduler())

        # Custom startup logic
        await self.on_startup()

        self.logger.info(f"{self.service_name} started successfully")

    async def stop(self):
        """Stop the service gracefully"""
        self.logger.info(f"Stopping {self.service_name}")

        # Custom shutdown logic
        await self.on_shutdown()

        # Close connections
        await self.db.disconnect()
        await self.cache.disconnect()

        self.logger.info(f"{self.service_name} stopped")

    async def _health_check_scheduler(self):
        """Periodic health check scheduler"""
        while True:
            try:
                await self.health_check()
                await asyncio.sleep(self.config.health_check_interval)
            except Exception as e:
                self.logger.error(f"Health check scheduler error: {str(e)}")
                await asyncio.sleep(self.config.health_check_interval)

    @abstractmethod
    async def on_startup(self):
        """Custom startup logic - implement in child classes"""
        pass

    @abstractmethod
    async def on_shutdown(self):
        """Custom shutdown logic - implement in child classes"""
        pass